docs/grammar/SolidityParser.g4

```diff
@@ -172,19 +172,27 @@ locals[
 		| modifierInvocation
 		| {!$virtualSet}? Virtual {$virtualSet = true;}
 		| {!$overrideSpecifierSet}? overrideSpecifier {$overrideSpecifierSet = true;}
-	 )*
+	)*
 	(Returns LParen returnParameters=parameterList RParen)?
 	(Semicolon | body=block);
 
 /**
  * The definition of a free function.
  */
- freeFunctionDefinition:
- 	Function (identifier | Fallback | Receive)
- 	LParen (arguments=parameterList)? RParen
- 	stateMutability?
- 	(Returns LParen returnParameters=parameterList RParen)?
- 	(Semicolon | body=block);
+freeFunctionDefinition
+locals[
+	boolean mutabilitySet = false,
+	boolean suffixSet = false,
+]
+:
+	Function (identifier | Fallback | Receive)
+	LParen (arguments=parameterList)? RParen
+	(
+		{!$mutabilitySet}? stateMutability {$mutabilitySet = true;}
+		| {!$suffixSet}? visibility {$suffixSet = true;}
```

Comment:
visibility?

Also I'm not entirely convinced that in this version it's really that justified to split it out from contract function definitions... I'd prefer to have one function definition node that can be free or not and just restrict modifiers and suffix accordingly - otherwise this has a lot of duplication really and does not actually match our AST structure (which again we don't have to be *draconian* in following, but if possible, it's preferable)

Reply:
Damn :) But yeah tests caught this as expected. I just did not get to look at them yet.

> Also I'm not entirely convinced that in this version it's really that justified to split it out from contract function definitions...

You had a lot of good points about it earlier (i.e. solc parser does not distinguish them and I ended up having to switch back to the flags anyway) so I agree. It ended up being more duplication than I thought it would be originally.

I can merge them back. Do you prefer it here or in a separate PR?

Reply:
Ah, wait. The original reason was also that issue with modifiers being repeatable but `suffix` not. I can merge it but I guess then we'll have two levels of flags? Won't that be more complicated?

Reply:
> I'd prefer to have one function definition node that can be free or not and just restrict modifiers and suffix accordingly

I tried to do that just now, but I suspect it may and up being pretty complex, unless ANTLR has some special feature for cases like this.

Basically, the issue is that you have to know whether you're dealing with a contract function or a free function to say if something like `suffix suffix` is allowed on the modifier list. Separating them into two definitions achieves that pretty nicely. If they're combined, I need to have a way to tell which one I'm dealing with:
- If rules could be parameterized, I would make that a parameter, but AFAIK ANTLR does not support that.
- I could have a variable set before entering the rule and cleared after exiting. But a stateful rule does not seem like a good thing (if it's even possible in the first place, which I suspect it's not).
- I could make it conditional on the presence of a visibility modifier, because that's the only thing that is always present on contract functions and never present on free functions. But since modifiers can be in any order, it seems like achieving this may require an overcomplicated rule.

Do you have an idea on how to do it cleanly?

Reply:
Look at ``elementaryTypeName`` - it should be possible to handle "being a free function" similarly to "allowAddressPayable" for that one. So that's pretty much exactly what you need here - and basically means splitting the rule into a free function and a contract function version, just without duplicating it :-). I mentioned that in some other comment, but not sure which one, must have gotten lost there :-).

Reply:
I.e. you can just make modifiers, resp. suffices (that is whatever differs) conditional on that additional flag of being free and share the rest - and on use of functions just specify whether they're free or not (like e.g. in ``elementaryTypeName`` occurring in ``typeName`` vs. occurring in ``expression``)

Reply:
So that's option one. Looks like it's possible after all. Yeah, that's the best option.

I think I saw you mentioning a solution for this before, but it wasn't where I thought it was when I tried to find it later. When I looked for it I saw answers that ANTLR does not allow parametric rules.

Reply:
Ah, yeah, nice - you agree that this is much nicer and less duplicated than the split now, right :-)? And also less of a relevant change for anyone actually using the grammar (I'd still be curious if anybody actually properly does or whether it's just people merely experimenting a bit with it here and there). Looks good now in any case.

Reply:
Yes, I agree with the knowledge I have now. I just initially thought it was not possible and earlier I missed some nuances which led me to think that a split matched our parser better.